tests/compiletest.rs

```diff
@@ -1,13 +1,35 @@
 use colored::*;
 use regex::Regex;
 use std::path::{Path, PathBuf};
-use std::{env, ffi::OsString};
+use std::{env, ffi::OsString, process::Command};
 use ui_test::{color_eyre::Result, Config, DependencyBuilder, Mode, OutputConflictHandling};
 
 fn miri_path() -> PathBuf {
     PathBuf::from(option_env!("MIRI").unwrap_or(env!("CARGO_BIN_EXE_miri")))
 }
 
+// Build the shared object file for testing external C function calls.
+fn build_so_for_c_ffi_tests() {
+    let cc = option_env!("CC").unwrap_or("cc");
+    let cc_output = Command::new(cc)
+        .args([
+            "-shared",
+            "-o",
+            "tests/extern-so/libtestlib.so",
+            "tests/extern-so/test.c",
```

Comment:
We need to get this to work in `x.py` and on Rust CI. One issue there is that the source directory will be read-only. So we need to do this build somewhere in the target dir. There is a CARGO_TARGET_DIR env var we can probably use for that. Use some 'miri-extern-so' subfolder of CARGO_TARGET_DIR.